test/functional/wallet_reorgsrestore.py

```diff
@@ -90,6 +91,48 @@ def test_coinbase_automatic_abandon_during_startup(self):
         # Verify the coinbase descendant was also marked as abandoned
         assert_equal(wallet0.gettransaction(descendant_tx_id)['details'][0]['abandoned'], True)
 
+    def test_reorg_handling_during_unclean_shutdown(self):
+        self.log.info("Test that wallet doesn't crash due to a duplicate block disconnection event after an unclean shutdown")
+        node = self.nodes[0]
+        # Receive coinbase reward on a new wallet
+        node.createwallet(wallet_name="reorg_crash", load_on_startup=True)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        self.generatetoaddress(node, 1, wallet.getnewaddress(), sync_fun=self.no_op)
+
+        # Restart to ensure node and wallet are flushed
+        self.restart_node(0)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        assert_greater_than(wallet.getwalletinfo()['immature_balance'], 0)
+
+        # Disconnect tip and sync wallet state
+        tip = wallet.getbestblockhash()
+        wallet.invalidateblock(tip)
+        wallet.syncwithvalidationinterfacequeue()
+
+        # Tip was disconnected, ensure coinbase has been abandoned
+        assert_equal(wallet.getwalletinfo()['immature_balance'], 0)
+        coinbase_tx_id = wallet.getblock(tip, verbose=1)["tx"][0]
+        assert_equal(wallet.gettransaction(coinbase_tx_id)['details'][0]['abandoned'], True)
+
+        # Abort process abruptly to mimic an unclean shutdown (no flush)
+        # The best block locator changes might not have been persisted
+        node.process.kill()
+
+        # Start wallet again. Ensure the tx is still abandoned and the wallet has no balance
+        self.start_node(0)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        assert_equal(wallet.getbestblockhash(), tip)
```

Comment:
Re: https://github.com/bitcoin/bitcoin/pull/31757#discussion_r1961637704
Continuing this thread here because there doesn't seem to be a way to continue that thread while making it part of the review.

As per this comment and related code https://github.com/bitcoin/bitcoin/issues/31824#issuecomment-2669011985, I understand that `FlushStateMode::IF_NEEDED` mode is used and therefore I assume that makes it quite unlikely to flush chainstate in a functional test, so the test will not fail.

> The test will still pass if the best locator is flushed to disk before the abrupt shutdown. And that’s fine, that’s the expected behavior.

However, I don't understand this comment. Hypothetically, if the state were to be flushed before abrupt node kill, wouldn't this assertion here fail?

Reply:
> However, I don't understand this comment. Hypothetically, if the state were to be flushed before abrupt node kill, wouldn't this assertion here fail?

I was talking about the wallet best locator, which is different to the node's best block (`getwalletinfo()['lastprocessedblock']` vs `getbestblockhash()`). 
Just updated the test to reflect this in a more understandable manner.